source/extensions/clusters/dns_srv/dns_srv_cluster.cc

```diff
@@ -0,0 +1,351 @@
+              // SRV record target is an IP address, not a hostname.
+              ResolveTargetPtr target = ResolveTargetPtr(
+                  new ResolveTarget(*active_resolve_list_, dns_resolver_, dns_lookup_family_,
+                                    dns.srv().target_, dns.srv().port_));
+
+              active_resolve_list_->addResolvedTarget(std::move(target), address);
+            } else {
+              active_resolve_list_->addTarget(ResolveTargetPtr(
+                  new ResolveTarget(*active_resolve_list_, dns_resolver_, dns_lookup_family_,
+                                    dns.srv().target_, dns.srv().port_)));
+            }
+          }
+
+          active_resolve_list_->noMoreTargets();
+        } else {
+          info_->configUpdateStats().update_failure_.inc();
+          resolve_timer_->enableTimer(dns_refresh_rate_ms_);
+          onPreInitComplete();
+        }
+      });
+}
+
+void DnsSrvCluster::allTargetsResolved() {
+  ASSERT(active_resolve_list_.get() != nullptr);
+
+  HostVector new_hosts;
+  absl::flat_hash_set<std::string> all_new_hosts;
+
+  const auto& locality_lb_endpoints = load_assignment_.endpoints()[0];
+  const auto& lb_endpoint_ = load_assignment_.endpoints()[0].lb_endpoints()[0];
+
+  PriorityStateManager priority_state_manager(*this, local_info_, nullptr, random_);
+  priority_state_manager.initializePriorityFor(locality_lb_endpoints);
+
+  // Only update the list of targets if at least one of them was resolved successfully
+  // this has a down-side: we can loose some of targets for some while but hopefully they'll get
+  // back next time
+  bool some_targets_resolved = false;
+
+  for (const auto& target : active_resolve_list_->getResolvedTargets()) {
+    // SRV query returns a number of instances (hostnames), but each hostname
+    // may potentially be resolved in a number of IP addresses
+    if (target->resolve_status_ != Network::DnsResolver::ResolutionStatus::Completed) {
+      ENVOY_LOG(debug, "IP resolution for target: '{}' has failed: {}",
+                target->srv_record_hostname_, target->resolve_status_details_);
+      continue;
+    }
+
+    some_targets_resolved = true;
+
+    for (auto& address : target->resolved_targets_) {
+      if (all_new_hosts.count(address->asString()) > 0) {
+        continue;
+      }
+
+      ENVOY_LOG(debug, "Endpoints size: {}", load_assignment_.endpoints().size());
+
+      auto host_or_status = HostImpl::create(
+          info_, target->srv_record_hostname_, address,
+          // TODO(zyfjeff): Created through metadata shared pool
+          std::make_shared<const envoy::config::core::v3::Metadata>(lb_endpoint_.metadata()),
+          std::make_shared<const envoy::config::core::v3::Metadata>(
+              locality_lb_endpoints.metadata()),
+          lb_endpoint_.load_balancing_weight().value(), locality_lb_endpoints.locality(),
+          lb_endpoint_.endpoint().health_check_config(), locality_lb_endpoints.priority(),
```

Comment:
Where did the values for priority and weight come from? These should be set from the srv response, right?